tests/analyser/analyser.cpp

```diff
@@ -389,24 +434,553 @@ TEST(Analyser, unsuitablyConstrained)
+
+    analyser->addExternalVariable(externalVariable);
+
+    EXPECT_TRUE(analyser->containsExternalVariable(externalVariable));
+    EXPECT_FALSE(analyser->containsExternalVariable(nullptr));
+}
+
+TEST(Analyser, externalVariableByIndex)
+{
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("generator/hodgkin_huxley_squid_axon_model_1952/model.cellml"));
+
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    auto analyser = libcellml::Analyser::create();
+
+    EXPECT_EQ(nullptr, analyser->externalVariable(0));
+
+    auto externalVariable = libcellml::AnalyserExternalVariable::create(model->component("membrane")->variable("V"));
+
+    analyser->addExternalVariable(externalVariable);
+
+    EXPECT_EQ(externalVariable, analyser->externalVariable(0));
+    EXPECT_EQ(nullptr, analyser->externalVariable(1));
+}
+
+TEST(Analyser, externalVariableByName)
+{
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("generator/hodgkin_huxley_squid_axon_model_1952/model.cellml"));
+
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    auto analyser = libcellml::Analyser::create();
+
+    EXPECT_EQ(nullptr, analyser->externalVariable(model, "membrane", "V"));
+
+    auto externalVariable = libcellml::AnalyserExternalVariable::create(model->component("membrane")->variable("V"));
+
+    analyser->addExternalVariable(externalVariable);
+
+    EXPECT_EQ(externalVariable, analyser->externalVariable(model, "membrane", "V"));
+    EXPECT_EQ(nullptr, analyser->externalVariable(model, "membrane", "X"));
+}
+
+TEST(Analyser, onePrimaryVoiExternalVariable)
+{
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("generator/hodgkin_huxley_squid_axon_model_1952/model.cellml"));
+
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    const std::vector<std::string> expectedIssues = {
+        "Variable 'time' in component 'environment' is marked as an external variable, but it is the variable of integration which cannot be used as an external variable.",
+    };
+
+    auto analyser = libcellml::Analyser::create();
+
+    analyser->addExternalVariable(libcellml::AnalyserExternalVariable::create(model->component("environment")->variable("time")));
+
+    analyser->analyseModel(model);
+
+    EXPECT_EQ_ISSUES(expectedIssues, analyser);
+}
+
+TEST(Analyser, oneNonPrimaryVoiExternalVariable)
+{
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("generator/hodgkin_huxley_squid_axon_model_1952/model.cellml"));
+
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    const std::vector<std::string> expectedIssues = {
+        "Variable 'time' in component 'membrane' is marked as an external variable, but it is equivalent to the primary variable of integration which cannot be used as an external variable.",
+    };
```

Comment:
I don't know what the "primary variable of integration" is?

Reply:
Done in commit 748b17092.